aptos-move/move-examples/marketplace/sources/listing.move

```diff
@@ -0,0 +1,259 @@
+/// Defines a single listing or an item for sale or auction. This is an escrow service that
+/// enables two parties to exchange one asset for another.
+/// Each listing has the following properties:
+/// * FeeSchedule specifying payment flows
+/// * Owner or the person that can end the sale or auction
+/// * Starting time
+/// * Logic for cleanup
+module marketplace::listing {
+    use std::error;
+    use std::option;
+    use std::signer;
+    use std::string::String;
+
+    use aptos_framework::object::{Self, ConstructorRef, DeleteRef, ExtendRef, Object, ObjectCore, TransferRef};
+    use aptos_framework::timestamp;
+
+    use aptos_token::token::{Self as tokenv1, Token as TokenV1};
+    use aptos_token_objects::token;
+    use aptos_token_objects::royalty;
+
+    use marketplace::fee_schedule::FeeSchedule;
+
+    friend marketplace::coin_listing;
+
+    /// There exists no listing.
+    const ENO_LISTING: u64 = 1;
+    /// The listing is not yet live.
+    const ELISTING_NOT_STARTED: u64 = 2;
+    /// The entity is not the creator.
+    const ENOT_CREATOR: u64 = 3;
+
+    // Core data structures
+
+    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
+    /// Corner-stone for all listings, represents the core utility layer including object
+    /// cleanup.
+    struct Listing has key {
+        /// The item owned by this listing, transferred to the new owner at the end.
+        object: Object<ObjectCore>,
+        /// The seller of the object
+        seller: address,
+        /// The fees associated with claiming this listing.
+        fee_schedule: Object<FeeSchedule>,
+        /// The Unix timestamp in seconds at which point bidding and purchasing can occur
+        start_time: u64,
+        /// Used to clean-up at the end.
+        delete_ref: DeleteRef,
+        /// Used to create a signer to transfer the listed item, ideally the TransferRef would
+        /// support this.
+        extend_ref: ExtendRef,
+    }
+
+    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
+    /// Contains a tokenv1 as an object
+    struct TokenV1Container has key {
+        /// The stored token.
+        token: TokenV1,
+        /// Used to cleanup the object at the end
+        delete_ref: DeleteRef,
+        /// Used to transfer the tokenv1 at the conclusion of a purchase.
+        transfer_ref: TransferRef,
+    }
+
+    // Init functions
+
+    public(friend) fun init(
+        creator: &signer,
+        object: Object<ObjectCore>,
```

Comment:
Shouldn't this be a token? Object < Token >. This could be generalized to object marketplace but other functionalities in this example are token-related anyway so having a general object input here is rather confusing IMO

Reply:
the original idea was to make this general, but agreed...

Reply:
actually, recall this takes the tokenv1 container...